server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1447,7 +1469,8 @@ private void ensureWriteAllowed(Engine.Operation.Origin origin) throws IllegalIn
         IndexShardState state = this.state; // one time volatile read
 
         if (origin.isRecovery()) {
-            if (state != IndexShardState.RECOVERING) {
+            final boolean isResetting = getEngineOrNull() instanceof SearchOnlyEngine;
```

Comment:
I am not a super fan of this. I wonder if we can afterwards rethink this.

Reply:
same here.
I would prefer to have a separate `runTranslogRecovery` method that we use for the local replay after reset. This method could then use a different stats object as well as a different origin.